File: packages/cli/src/types/entities.ts
/*
 Kibi — repo-local, per-branch, queryable long-term memory for software projects
 Copyright (C) 2026 Piotr Franczyk

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Affero General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Affero General Public License for more details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import type { ProofBinding, ProofContract } from "../public/proof-protocol.js";
import type { ProofReceipt } from "../public/proof-receipt.js";

export interface BaseEntity {
  id: string;
  title: string;
  status: string;
  created_at: string; // ISO 8601
  updated_at: string; // ISO 8601
  source: string; // URI
  tags?: string[];
  owner?: string;
  priority?: string;
  severity?: string;
  links?: string[];
  text_ref?: string;
}

// Typed fact fields per proposal
export interface FactFields {
  fact_kind?:
    | "subject"
    | "property_value"
    | "observation"
    | "meta"
    | "predicate_schema"
    | "predicate";
  subject_key?: string;
  property_key?: string;
  operator?: "eq" | "neq" | "lt" | "lte" | "gt" | "gte";
  value_type?: "string" | "int" | "number" | "bool";
  value_string?: string;
  value_int?: number;
  value_number?: number;
  value_bool?: boolean;
  unit?: string;
  scope?: string;
  polarity?: "require" | "forbid" | "assert" | "deny";
  closed_world?: boolean;
  valid_from?: string;
  valid_to?: string;
  canonical_key?: string;
  claim_key?: string;
  claim_text?: string;
  predicate_name?: string;
  predicate_namespace?: string;
  predicate_arity?: number;
  argument_names?: string[];
  argument_types?: string[];
  argument_descriptions?: string[];
  aliases?: string[];
  examples?: string[];
  predicate_args?: string[];
}

export interface TestVerificationFields {
  verification_scope?: "unit" | "integration" | "end_to_end";
  verification_perspective?: "internal" | "consumer";
  proof_contract?: ProofContract;
  proof_bindings?: readonly ProofBinding[];
  proof_receipts?: readonly ProofReceipt[];
}

export type Requirement = BaseEntity & {
  type: "req";
  semantic_text?: string;
  logic_claims?: string[];
};
export type Scenario = BaseEntity & { type: "scenario" };
export type TestEntity = BaseEntity & TestVerificationFields & { type: "test" };
export type ADR = BaseEntity & { type: "adr" };
export type Flag = BaseEntity & { type: "flag" };
export type Event = BaseEntity & { type: "event" };
export type Symbol = BaseEntity & { type: "symbol" };
export type Fact = BaseEntity & FactFields & { type: "fact" };

export type Entity =
  | Requirement
  | Scenario
  | TestEntity
  | ADR
  | Flag
  | Event
  | Symbol
  | Fact;

File: packages/cli/schema/entities.pl
% Module: kibi_entities
% Entity type and property definitions for Kibi knowledge base
:- module(kibi_entities, [entity_type/1, entity_property/3, required_property/2, optional_property/2]).

% Entity types
entity_type(req).
entity_type(scenario).
entity_type(test).
entity_type(adr).
entity_type(flag).
entity_type(event).
entity_type(symbol).
entity_type(fact).

% entity_property(EntityType, Property, Type).
% Basic typing hints (atom, string, datetime, list, uri)
entity_property(_, id, atom).
entity_property(_, title, string).
entity_property(_, status, atom).
entity_property(_, created_at, datetime).
entity_property(_, updated_at, datetime).
entity_property(_, source, uri).

% Optional properties
entity_property(_, tags, list).
entity_property(_, owner, atom).
entity_property(_, priority, atom).
entity_property(_, severity, atom).
entity_property(_, links, list).
entity_property(_, text_ref, uri).

% Required properties for all entity types
required_property(Type, id) :- entity_type(Type).
required_property(Type, title) :- entity_type(Type).
required_property(Type, status) :- entity_type(Type).
required_property(Type, created_at) :- entity_type(Type).
required_property(Type, updated_at) :- entity_type(Type).
required_property(Type, source) :- entity_type(Type).

% Optional properties for all entity types
optional_property(Type, tags) :- entity_type(Type).
optional_property(Type, owner) :- entity_type(Type).
optional_property(Type, priority) :- entity_type(Type).
optional_property(Type, severity) :- entity_type(Type).
optional_property(Type, links) :- entity_type(Type).
optional_property(Type, text_ref) :- entity_type(Type).

% Documentation helpers
% list all entity types
all_entity_types(Ts) :- findall(T, entity_type(T), Ts).
